Feature/facilitate component updates

jungeunyoon opened the pull request.

jungeunyoon commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Bridge classes - update components [3]](https://trello.com/c/9lZO3mKH/4634-bridge-classes-update-components-3)

rodrigopavezi commented on src/api/twitter.js:
```diff
@@ -2,12 +2,19 @@ const logger = require("../../src/logger");
+
 function closeStream() {
-  return stream.destroy();
+  if (stream) {return stream.destroy();}
```
@jungeunyoon I think we will need to add the streams created to a map with the component id as key online 39. and close them as per component id passed to the close function. Thoughts? cheers

jungeunyoon replied:
@rodrigopavezi Yes, great that you brought that up! I was considering different options for closing streams, which is why I brought up needing to accept messages for deleting components last week. 

Do you think it's better to to keep this in the components object that already exists or in a separate object for componentId -> stream? I'm leaning towards the latter to keep a separation and you seem to be suggesting the same.

rodrigopavezi replied:
@jungeunyoon I would simplify and create a streams object on the twitter.js which has company id as key for each new stream created. What do you think? cheers

rodrigopavezi commented on src/messaging/broadcast-ipc.js:
```diff
@@ -13,7 +13,7 @@ function twitterUpdate(data = {}) {
 
   const messageObject = Object.assign({}, data);
-  broadcast("TWITTER-UPDATE", messageObject)
+  broadcast("Twitter-Update", messageObject)
```
@jungeunyoon curious, what is the reason for it to be lower case?

jungeunyoon replied:
@rodrigopavezi Ah it should just be all lowercase so I'll fix that, thanks! My mistake on the Pascal Case. I think I was at some point considering writing tests for unusual casing. But back to your question, it's a unofficial standard I've noticed with all message topics being lowercase. Figured we should stick with that.

rodrigopavezi commented on src/components/components-controller.js:
```diff
@@ -0,0 +1,49 @@
+
+  if (!twitter.credentialsExist()) {return logger.file("Credentials do not exist - can not update components");}
+
+  twitter.closeStream();
```
@jungeunyoon I think here that we need to pass the component id for closing the specific stream. Thoughts? Cheers

jungeunyoon commented:
@rodrigopavezi This is ready for re-review, thanks! I'll squash commits after.

rodrigopavezi commented on src/messaging/messaging.js:
```diff
@@ -3,10 +3,18 @@ const commonConfig = require("common-display-module");
 
 function handleComponent(message) {
-  return update.process(message);
+  switch (message.status.toUpperCase()) {
```
@jungeunyoon I am trying to understand the scenario for deleting a component from the list of components. I can't see a component being able to send a deletion message. User will remove a widget with a component from the presentation. When presentation updates and reloads on player that removed component will not be anymore on the presentation. So I think instead of waiting for the delete status on the watch message the module should check if there are non active components and then delete them stopping the streaming. Thoughts?  Cheers

jungeunyoon commented on package.json:
```diff
@@ -7,9 +7,9 @@
-    "test": "eslint test && eslint src && mocha -r test/mocha-env.js -t 20000 --recursive --exit test/integration/** && mocha -r test/mocha-env.js -t 20000 --recursive --exit test/unit/**",
-    "test-unit": "eslint test && eslint src && mocha -r test/mocha-env.js -t 20000 --recursive --exit test/unit/**",
-    "test-integration": "eslint test && eslint src && mocha -r test/mocha-env.js -t 20000 --recursive --exit test/integration/**",
+    "test": "eslint test && eslint src && mocha -r test/mocha-env.js -t 30000 --recursive --exit test/integration/** && mocha -r test/mocha-env.js -t 20000 --recursive --exit test/unit/**",
```
@Rise-Vision/creative Increased the timeout since Twitter Stream tests seem to need more time 1/5 times.

jungeunyoon commented:
@rodrigopavezi Ready for review, thanks. 

@Rise-Vision/creative As per our discussion outside of Github, removed all calls to delete components / close streams outside of tests and added two cards --> https://trello.com/c/4IeMuw4Z/4672-delete-components-after-checking-stable-unstable-status-3
https://trello.com/c/HtyRoiZu/4671-add-component-to-lms-client-list-2

rodrigopavezi commented:
@jungeunyoon looks good 👍  cheers
